js/binance.js

```diff
@@ -741,22 +746,10 @@ module.exports = class binance extends Exchange {
                     throw new InsufficientFunds (this.id + ' ' + this.json (response));
                 } else if (error == -2011) {
                     throw new OrderNotFound (this.id + ' ' + this.json (response));
+                } else if (error < 0) {
+                    throw new ExchangeError (this.id + ' ' + this.json (response));
                 }
             }
         }
     }
-
-    async request (path, api = 'public', method = 'GET', params = {}, headers = undefined, body = undefined) {
-        let response = await this.fetch2 (path, api, method, params, headers, body);
-        if ('code' in response) {
-            if (response['code'] < 0) {
-                if (response['code'] == -2010)
-                    throw new InsufficientFunds (this.id + ' ' + this.json (response));
-                if (response['code'] == -2011)
-                    throw new OrderNotFound (this.id + ' ' + this.json (response));
-                throw new ExchangeError (this.id + ' ' + this.json (response));
-            }
-        }
-        return response;
-    }
```

Comment:
Hey ) Good to hear from you again ) Are you sure we can delete the request method? I mean, the can return HTTP 200 + JSON error code...

Happy holidays!)

Reply:
A possible way to solve this is to define some throwExceptionOnErrorCode (code, reply, .. ? ) function and move it all their to get rid of the logic duplication.

Reply:
Hi!

I'd say that 200 + error code is somewhat a strange couple, at least by reading their doc.

Actually, it boils down to error handling strategy.
I used to open the gates wide and then catch and handle unseen errors as they appear.
Another approach would be to catch everything suspicious with generic ExchangeErorr and then narrow it down.

So if you want me to put 200+error hadling back just in case then no problem, I'll do that.

Reply:
Could you unify the logic under that if-conditional from handleErrors() and the same logic from request() into a single method and call it then from both places? I'd say that would be the best way for us. Yes, from their docs it looks like 200 + error is impossible, but, in fact, there's no strict rule for that. We already had those errors, otherwise I would not duplicate it, of course... That duplication in itself was an example of bad coding anyway, I admit )

Reply:
Actually, I've been writing a test suite with the goal to check exchange trading and error handling logic (which is missing from ccxt test cases).

In particular, it creates and cancels open orders, creates orders that throw InsufficientFunds, cancels executed orders (that's why I implemented fetchClosedOrders for binance) etc. This is a user testing, something that you don't want to do on developer's side as it would require holding balances on plenty of exchanges.

I submit patches once they pass this testsuite so I'm quite confident we're on the safe side (being more precise, I create patches once existing implementation starts to break this testsuite which I run regularly).

P.S. Just in case you wanted to ask: I'll submit a PR for this testsuite eventually (once it's generic enough) so everyone on the user side can be confident that trading is safe.

Reply:
> Just in case you wanted to ask: I'll submit a PR for this testsuite eventually (once it's generic enough) so everyone on the user side can be confident that trading is safe.

That would be awesome, man!

Reply:
So, reverting back to your question if I'm sure we can delete `request`.

Personally, I'm sure as this patch passed the testsuite and it works.

But if you know that for some reason sometimes binance returns exceptions with `200` then I'll have to redo that.

Reply:
I'd say this edit would make us sleep well )

From this:

```JavaScript

    handleErrors (code, reason, url, method, headers, body) {
        if (code >= 400) {
            if (code == 418)
                throw new DDoSProtection (this.id + ' ' + code.toString () + ' ' + reason + ' ' + body);
            if (body.indexOf ('MIN_NOTIONAL') >= 0)
                throw new InvalidOrder (this.id + ' order cost = amount * price should be > (0.001 BTC or 0.01 ETH or 1 BNB or 1 USDT)' + body);
            if (body.indexOf ('LOT_SIZE') >= 0)
                throw new InvalidOrder (this.id + ' order amount should be evenly divisible by lot size, use this.amountToLots (symbol, amount) ' + body);
            if (body.indexOf ('PRICE_FILTER') >= 0)
                throw new InvalidOrder (this.id + ' order price exceeds allowed price precision or invalid, use this.priceToPrecision (symbol, amount) ' + body);
            if (body.indexOf ('Order does not exist') >= 0)
                throw new OrderNotFound (this.id + ' ' + body);
            if (body[0] == "{") {
                let response = JSON.parse (body);
                let error = this.safeValue (response, 'code');
                if (error == -2010) {
                    throw new InsufficientFunds (this.id + ' ' + this.json (response));
                } else if (error == -2011) {
                    throw new OrderNotFound (this.id + ' ' + this.json (response));
                }
            }
        }
    }

```

↓↓↓ To this ↓↓↓

```JavaScript
    handleErrors (code, reason, url, method, headers, body) {
        if (body[0] == "{") {
            let response = JSON.parse (body);
            let error = this.safeValue (response, 'code');
            if (error == -2010) {
                throw new InsufficientFunds (this.id + ' ' + this.json (response));
            } else if (error == -2011) {
                throw new OrderNotFound (this.id + ' ' + this.json (response));
            }
        }
        if (code >= 400) {
            if (code == 418)
                throw new DDoSProtection (this.id + ' ' + code.toString () + ' ' + reason + ' ' + body);
            if (body.indexOf ('MIN_NOTIONAL') >= 0)
                throw new InvalidOrder (this.id + ' order cost = amount * price should be > (0.001 BTC or 0.01 ETH or 1 BNB or 1 USDT)' + body);
            if (body.indexOf ('LOT_SIZE') >= 0)
                throw new InvalidOrder (this.id + ' order amount should be evenly divisible by lot size, use this.amountToLots (symbol, amount) ' + body);
            if (body.indexOf ('PRICE_FILTER') >= 0)
                throw new InvalidOrder (this.id + ' order price exceeds allowed price precision or invalid, use this.priceToPrecision (symbol, amount) ' + body);
            if (body.indexOf ('Order does not exist') >= 0)
                throw new OrderNotFound (this.id + ' ' + body);
        }
    }
```

And we can remove the request() method then. What do you think?

Reply:
Yeah, I came up with the same idea too )